modules/SLAPEnrich.py

```diff
@@ -1,287 +1,117 @@
-                                is_associated=True,
-                                #TODO check is this the correct evidence code "computational combinatorial evidence"
-                                evidence_codes=["http://purl.obolibrary.org/obo/ECO_0000053"],
-                                provenance_type=provenance_type,
-                                resource_score=resource_score,
-                                urls=[linkout]
-                            )
-
-                            # *** Build unique_association_field object ***
-                            evidenceString.unique_association_fields = {
-                                'target_id': evidenceString.target.id,
-                                'pathway_id': 'http://www.reactome.org/PathwayBrowser/#%s' % (pathway_id),
-                                'disease_id': evidenceString.disease.id
-                            }
-
-                            error = evidenceString.validate(logging)
-
-                            if error > 0:
-                                self.logger.error(evidenceString.to_JSON())
-                                sys.exit(1)
-
-                            self.evidence_strings.append(evidenceString)
-
-                        else:
-                            self.logger.error("%s is not found in Ensembl" % gene_symbol)
-
-            self.logger.info("%s evidence parsed"%(n-1))
-            self.logger.info("%s evidence created"%len(self.evidence_strings))
-
-        slapenrich_input.close()
-
-    def write_evidence(self, filename=Config.SLAPENRICH_EVIDENCE_FILENAME):
-        self.logger.info("Writing SLAPEnrich evidence strings")
-        with open(filename, 'w') as slapenrich_output:
-            n = 0
-            for evidence_string in self.evidence_strings:
-                n += 1
-                self.logger.info(evidence_string.disease.id[0])
-                # get max_phase_for_all_diseases
-                error = evidence_string.validate(logging)
-                if error == 0:
-                    slapenrich_output.write(evidence_string.to_JSON(indentation=None)+"\n")
-                else:
-                    self.logger.error("REPORTING ERROR %i" %n)
-                    self.logger.error(evidence_string.to_JSON(indentation=4))
-            slapenrich_output.close()
 
+        return self.dataframe
+
+    @staticmethod
+    def parseEvidenceString(row):
+        try:
+            evidence = {
+                "datasourceId" : "slapenrich",
+                "datatypeId" : "affected_pathway",
+                "diseaseFromSource" : row["Cancer_type_acronym"],
+                "diseaseFromSourceMappedId" if row["EFO_id"] else None : row["EFO_id"],
+                "resourceScore" : row["pval"],
+                "pathwayName" : row["pathwayDescription"],
+                "pathwayId" : row["pathwayId"],
+                "targetFromSourceId" : row["gene"]
+            }
+            evidence.pop(None, None)
```

Comment:
What does this pop do?

Reply:
I see: if `row["EFO_id"] is None` an extra key/value pair (`None:None`) is added to the dictionary. In which case you need to pop that key. But at least the `"diseaseFromSourceMappedId"` will not be there with the `None` value. 

I would consider doing:
```Python
evidence = {
    "datasourceId" : "slapenrich",
    "datatypeId" : "affected_pathway",
    "diseaseFromSource" : row["Cancer_type_acronym"],
    "resourceScore" : row["pval"],
    "pathwayName" : row["pathwayDescription"],
    "pathwayId" : row["pathwayId"],
    "targetFromSourceId" : row["gene"]
}
if row['EFO_id]:
   evidence['diseaseFromSourceMappedId'] = row['EFO_id]
```

But I think if you use pyspark's json export methods, keys with `None` values will not be written into the JSON.

Reply:
I like it but it would still fail when using `--mappingStep` because `row['EFO_id']` would not be defined, right? So peharps a `if 'EFO_id' in row:` is needed?

Reply:
I have to think about... the json generation should be the very last step, this decision should happen before. But I might misunderstood.